app/actions/websocket/channel.ts

```diff
@@ -1,21 +1,68 @@
 // Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
 // See LICENSE.txt for license information.
 
+import {Model} from '@nozbe/watermelondb';
 import {DeviceEventEmitter} from 'react-native';
 
-import {localRemoveUserFromChannel, localSetChannelDeleteAt, switchToChannel} from '@actions/local/channel';
-import {updateUsersNoLongerVisible} from '@actions/remote/user';
+import {localRemoveCurrentUserFromChannel, localSetChannelDeleteAt, switchToChannel} from '@actions/local/channel';
+import {fetchChannelInfo, fetchMyChannel} from '@actions/remote/channel';
+import {fetchUsersByIds, updateNewUsersVisible, updateUsersNoLongerVisible} from '@actions/remote/user';
 import Events from '@constants/events';
 import DatabaseManager from '@database/manager';
 import {queryActiveServer} from '@queries/app/servers';
 import {deleteChannelMembership, queryCurrentChannel} from '@queries/servers/channel';
-import {queryConfig, setCurrentChannelId} from '@queries/servers/system';
+import {queryCommonSystemValues, queryConfig, setCurrentChannelId} from '@queries/servers/system';
 import {queryLastChannelFromTeam} from '@queries/servers/team';
-import {queryCurrentUser} from '@queries/servers/user';
+import {queryCurrentUser, queryUserById} from '@queries/servers/user';
 import {dismissAllModals, popToRoot} from '@screens/navigation';
 import {isTablet} from '@utils/helpers';
 import {isGuest} from '@utils/user';
 
+export async function handleUserAddedEvent(serverUrl: string, msg: any) {
+    const database = DatabaseManager.serverDatabases[serverUrl];
+    if (!database) {
+        return;
+    }
+    const {currentChannelId, currentUserId} = await queryCommonSystemValues(database.database);
+    const teamId = msg.data.team_id;
+    const channelId = msg.data.channel_id;
+    const userId = msg.data.user_id;
+
+    const models: Model[] = [];
+
+    try {
+        // Handle new visible users for guests
+        const addedUser = queryUserById(database.database, userId);
+        if (!addedUser) {
+            const {users} = await fetchUsersByIds(serverUrl, [userId], true);
+            if (users) {
+                models.push(...await database.operator.handleUsers({users, prepareRecordsOnly: true}));
+            }
+        }
+
+        if (userId === currentUserId) {
+            const {channels, memberships} = await fetchMyChannel(serverUrl, teamId, channelId, true);
+            if (channels && memberships) {
+                models.push(...await database.operator.handleMyChannel({channels, myChannels: memberships, prepareRecordsOnly: true}));
+            }
+            const {users} = await updateNewUsersVisible(serverUrl, true);
+            if (users) {
+                models.push(...await database.operator.handleUsers({users, prepareRecordsOnly: true}));
+            }
```

Comment:
why do we need this ?

Reply:
This was mainly thought for guests, but might not be needed if the information is being fetched somewhere else. The reasoning is the following:

A guest is only able to see people that share a channel with them. So... the same way we remove users from the database when we leave a channel if they are no longer visible, we add new users to the database when we enter a channel.

What I am missing for sure here is a check on whether the user is a guest or not.

Reply:
That is true, now this is what I think.

- User is added to a channel (it can be a guest user)
- Channel is displayed in the channel list
- User taps on the channel
- Posts are loaded for that channel & all the authors of those posts
- User opens the `Create conversation` modal and he only sees people that he can talk to as this is what the server returns.

Now if the user is removed and is a guest, then yes.. to remove users from the database as well as the channel and posts make sense, and then I guess we should remove the DM's that match the users that got removed.

is my logic sound?

Reply:
I wonder where do we do the queries for all users. Here are the things that come to mind:
- Create new DM screen
- Add new people to the channel
- User Autocomplete
- Searchbar user Autocomplete
- Interactive Dialog / App Form user autocomplete

Do we expect all of those to always ask the server? Or do we expect them to look at the database?
If any of those is looking at the database, we should keep the database up to date both when added and removed.


Regarding DMs, any existing DM for guests grant them visibility to that person (you don't lose the ability to talk with someone you have already have a DM with if the channel membership changes).

Reply:
Dm's cool got it

We expect those places to use the database and also fetch newer data